File: Fitness-ai-frontend/src/types/pose-analysis.ts
export type PoseAnalysisStatus = "idle" | "done" | "failed";
export type PoseAnalysisJobStatus = "queued" | "running" | "succeeded" | "failed";

export interface PoseAnalysisModelMetadata {
  name: string | null;
  input_size: number | null;
}

export interface PoseAnalysisSummary {
  total_frames: number;
  processed_frames: number;
  sampled_frames: number;
  valid_frame_count: number;
  average_confidence: number;
  source_fps: number | null;
  sample_fps: number;
}

export interface PoseAnalysisKeypoint {
  name: string;
  x: number;
  y: number;
  score: number;
}

export interface PoseAnalysisFrame {
  frame_index: number;
  timestamp_ms: number;
  keypoints: PoseAnalysisKeypoint[];
}

export interface PoseAnalysisResult {
  record_id: number;
  schema_version: number;
  status: PoseAnalysisStatus;
  model: PoseAnalysisModelMetadata | null;
  summary: PoseAnalysisSummary | null;
  frames: PoseAnalysisFrame[];
  error: string | null;
}

export interface PoseAnalysisJob {
  id: number;
  record_id: number;
  status: PoseAnalysisJobStatus;
  error: string | null;
  result_summary: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  completed_at: string | null;
}

File: app/schemas/pose_analysis.py
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class PoseAnalysisTriggerRequest(BaseModel):
    sample_fps: Optional[int] = Field(None, ge=1, le=30)


class PoseAnalysisModelMetadata(BaseModel):
    backend: Optional[str] = None
    name: Optional[str] = None
    input_size: Optional[int] = None


class PoseAnalysisSummary(BaseModel):
    total_frames: int = 0
    processed_frames: int = 0
    sampled_frames: int = 0
    valid_frame_count: int = 0
    average_confidence: float = 0
    source_fps: Optional[float] = None
    sample_fps: int


class PoseAnalysisFrame(BaseModel):
    frame_index: int
    timestamp_ms: int
    coordinate_space: Optional[str] = None
    frame: Optional[Dict[str, Any]] = None
    keypoints: List[Dict[str, Any]]


class PoseAnalysisResponse(BaseModel):
    record_id: int
    video_revision: int = 0
    analysis_revision: Optional[int] = None
    schema_version: int = 1
    status: Literal["idle", "done", "failed"]
    model: Optional[PoseAnalysisModelMetadata] = None
    summary: Optional[PoseAnalysisSummary] = None
    frames: List[PoseAnalysisFrame] = []
    error: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


PoseAnalysisJobStatus = Literal["queued", "running", "succeeded", "failed", "cancelled"]


class PoseAnalysisJobResponse(BaseModel):
    id: int
    record_id: int
    video_revision: int
    sample_fps: Optional[int] = None
    status: PoseAnalysisJobStatus
    error: Optional[str] = None
    result_summary: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)
